File: ghost-env/src/ghost-env.ts
import type { ChaosOptions } from "./chaos.js";
import { applyChaos } from "./chaos.js";
import { Recorder, type CallRecord } from "./recording/recorder.js";
import { mulberry32 } from "./world/seed.js";
import { WorldState } from "./world/state.js";

export interface Provider {
  name: string;
  seed?: (world: WorldState, rng: () => number) => void;
  handle?: (url: URL, method: string, body: unknown, world: WorldState) => Promise<Response | null>;
}

export interface GhostEnvConfig {
  seed?: number;
  providers: Provider[];
  chaos?: ChaosOptions;
}

export class GhostEnv {
  readonly world = new WorldState();
  private readonly recorder = new Recorder();
  private readonly rng: () => number;
  private readonly providers: Provider[];
  private readonly chaos?: ChaosOptions;
  private snapshotStr: string | null = null;

  constructor(config: GhostEnvConfig) {
    this.providers = config.providers;
    this.chaos = config.chaos;
    this.rng = mulberry32(config.seed ?? 1);
    for (const p of this.providers) {
      p.seed?.(this.world, this.rng);
    }
  }

  reset(): void {
    if (this.snapshotStr == null) {
      this.recorder.clear();
      this.world.restore("{}");
      for (const p of this.providers) {
        p.seed?.(this.world, this.rng);
      }
      return;
    }
    this.world.restore(this.snapshotStr);
    this.recorder.clear();
  }

  snapshot(): string {
    this.snapshotStr = this.world.snapshot();
    return this.snapshotStr;
  }

  restore(data: string): void {
    this.world.restore(data);
  }

  async fetch(input: string | URL, init?: RequestInit): Promise<Response> {
    const url = typeof input === "string" ? new URL(input) : new URL(input.toString());
    const method = (init?.method ?? "GET").toUpperCase();
    let body: unknown;
    if (init?.body) {
      const t = typeof init.body === "string" ? init.body : String(init.body);
      try {
        body = JSON.parse(t);
      } catch {
        body = t;
      }
    }
    const start = performance.now();
    let matchedName = "unknown";
    const run = async () => {
      for (const p of this.providers) {
        if (!p.handle) continue;
        const res = await p.handle(url, method, body, this.world);
        if (res) {
          matchedName = p.name;
          return res;
        }
      }
      throw new Error(`ghost-env: no provider matched ${method} ${url}`);
    };
    try {
      const res = await applyChaos(this.chaos, this.rng, run);
      const text = await res.clone().text();
      let parsed: unknown = text;
      try {
        parsed = JSON.parse(text);
      } catch {
        /* keep string */
      }
      const rec: Omit<CallRecord, "id"> = {
        provider: matchedName,
        method,
        url: url.toString(),
        requestBody: body,
        responseStatus: res.status,
        responseBody: parsed,
        durationMs: performance.now() - start,
      };
      this.recorder.record(rec);
      return res;
    } catch (e) {
      this.recorder.record({
        provider: "error",
        method,
        url: url.toString(),
        requestBody: body,
        responseStatus: 500,
        responseBody: String(e),
        durationMs: performance.now() - start,
      });
      throw e;
    }
  }

  calls(provider?: string): CallRecord[] {
    return this.recorder.filter(provider);
  }

  wasCalled(
    provider: string,
    partial?: Partial<Pick<CallRecord, "method" | "url">> & { pathIncludes?: string },
  ): boolean {
    return this.calls(provider).some((c) => {
      if (partial?.method && c.method !== partial.method) return false;
      if (partial?.url && c.url !== partial.url) return false;
      if (partial?.pathIncludes && !c.url.includes(partial.pathIncludes)) return false;
      return true;
    });
  }

  /** Minimal SQL facade backed by world entities `pg:table`. */
  db(_name = "postgres") {
    const world = this.world;
    return {
      async query(sql: string, params: unknown[] = []) {
        const m = sql.match(/FROM\s+(\w+)/i);
        const table = m?.[1]?.toLowerCase() ?? "unknown";
        const rows = world.list(`pg:${table}`) as Record<string, unknown>[];
        const filtered =
          /WHERE\s+(\w+)\s*=\s*\$1/i.test(sql) && params.length
            ? rows.filter((r) => String(r[params[0] === "admin" ? "role" : Object.keys(r)[0]]) === String(params[0]))
            : rows;
        return { rows: filtered, rowCount: filtered.length };
      },
    };
  }
}

File: ghost-env/python/src/ghost_env/ghost_env.py
from __future__ import annotations

import json
import re
import time
from typing import Any, Callable, TypedDict

from ghost_env.recorder import Recorder
from ghost_env.seed import mulberry32
from ghost_env.world import WorldState


class Provider(TypedDict, total=False):
    name: str
    seed: Callable[[WorldState, Callable[[], float]], None]
    handle: Callable[[str, str, Any, WorldState], tuple[int, str] | None]


class GhostEnvConfig(TypedDict, total=False):
    seed: int
    providers: list[Provider]
    chaos: dict[str, Any]


class GhostEnv:
    def __init__(self, config: GhostEnvConfig) -> None:
        self.world = WorldState()
        self._recorder = Recorder()
        self._providers = config["providers"]
        self._chaos = config.get("chaos") or {}
        self._rng = mulberry32(int(config.get("seed", 1)))
        self._snap: str | None = None
        for p in self._providers:
            if p.get("seed"):
                p["seed"](self.world, self._rng)

    def reset(self) -> None:
        if self._snap is None:
            self._recorder.clear()
            self.world.restore("{}")
            for p in self._providers:
                if p.get("seed"):
                    p["seed"](self.world, self._rng)
            return
        self.world.restore(self._snap)
        self._recorder.clear()

    def snapshot(self) -> str:
        self._snap = self.world.snapshot()
        return self._snap

    def restore(self, data: str) -> None:
        self.world.restore(data)

    def fetch(self, url: str, method: str = "GET", body: str | None = None) -> tuple[int, str]:
        parsed_body: Any = None
        if body:
            try:
                parsed_body = json.loads(body)
            except json.JSONDecodeError:
                parsed_body = body
        start = time.perf_counter()
        matched = "unknown"

        lat = float(self._chaos.get("min_latency_ms", 0) or 0)
        if lat:
            time.sleep(lat / 1000.0)
        fr = float(self._chaos.get("failure_rate", 0) or 0)
        if fr and self._rng() < fr:
            self._recorder.record(
                provider="error",
                method=method,
                url=url,
                request_body=parsed_body,
                response_status=500,
                response_body="chaos",
                duration_ms=(time.perf_counter() - start) * 1000,
            )
            raise RuntimeError("ghost-env chaos")

        status = 500
        text = ""
        try:
            for p in self._providers:
                h = p.get("handle")
                if not h:
                    continue
                res = h(url, method.upper(), parsed_body, self.world)
                if res:
                    matched = p["name"]
                    status, text = res
                    break
            else:
                raise RuntimeError(f"ghost-env: no provider matched {method} {url}")
        except Exception as e:  # noqa: BLE001
            self._recorder.record(
                provider="error",
                method=method,
                url=url,
                request_body=parsed_body,
                response_status=500,
                response_body=str(e),
                duration_ms=(time.perf_counter() - start) * 1000,
            )
            raise

        duration = (time.perf_counter() - start) * 1000
        try:
            parsed_resp = json.loads(text)
        except json.JSONDecodeError:
            parsed_resp = text
        self._recorder.record(
            provider=matched,
            method=method,
            url=url,
            request_body=parsed_body,
            response_status=status,
            response_body=parsed_resp,
            duration_ms=duration,
        )
        return status, text

    def calls(self, provider: str | None = None):
        return self._recorder.filter(provider)

    def was_called(self, provider: str, **kwargs: Any) -> bool:
        method = kwargs.get("method")
        path_includes = kwargs.get("path_includes")
        for c in self.calls(provider):
            if method and c.method != method:
                continue
            if path_includes and path_includes not in c.url:
                continue
            return True
        return False

    def db(self, _name: str = "postgres"):
        world = self.world

        class Db:
            @staticmethod
            def query(sql: str, params: list[Any] | None = None):
                params = params or []
                m = re.search(r"FROM\s+(\w+)", sql, re.I)
                table = (m.group(1).lower() if m else "unknown")
                rows = world.list(f"pg:{table}")
                return {"rows": rows, "row_count": len(rows)}

        return Db()
